Added test and support for content type headers

joe-elliott opened the pull request:
- Added unit test and supporting class TestContentTypeMessageHandler to test the Content-Type header making it through the cache
- Added support to pass Content Headers through Cache by storing them in the CacheData object.

Commit:
Added test and support for content type headers

Signed-off-by: Joe Elliott <jelliott@qsrautomations.com>

wparad commented on test/Cimpress.Extensions.Http.Caching.InMemory.UnitTests/TestMessageHandler.cs:
```diff
@@ -9,9 +9,9 @@ internal class TestMessageHandler : HttpMessageHandler
-        private readonly HttpStatusCode responseStatusCode;
-        private readonly string content;
+        
+        protected readonly HttpStatusCode responseStatusCode;
```
Please find a different way to test this.  Creating protected methods is not the right approach.

wparad commented on test/Cimpress.Extensions.Http.Caching.InMemory.UnitTests/TestMessageHandler.cs:
```diff
@@ -26,13 +26,16 @@ protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage reques
+            return Task.FromResult(makeResponse());
+        }
+
+        protected virtual HttpResponseMessage makeResponse()
```
Same for here.

wparad commented on src/Cimpress.Extensions.Http.Caching.Abstractions/HttpContentHeadersExtensions.cs:
```diff
@@ -0,0 +1,26 @@
+    /// <summary>
+    /// Extension methods of the HttpResponseMessage that are related to the caching functionality.
+    /// </summary>
+    public static class HttpContentHeadersExtensions
```
Please make it internal.

wparad reviewed (changes requested):
It is really unfortunate that Microsoft forces us to do this hack, but I think it implemented the best way we can.  Please just improve the unit tests are I'll merge this in.

Commit:
Changed accessibility level of HttpContentHeadersExtensions to internal

joe-elliott commented:
Updated HttpContentHeadersExtensions to be internal.

Regarding the protected issue:  I often use protected/virtual to expose elements to child classes to customize behavior.  

Do you have a preferred pattern here?  I'd be glad to learn and implement something different.

wparad commented:
Composition instead of inheritance, but in this case I should say either add the features you need to the existing `TestMessageHandler`, i.e. another private field for encoding, etc... or copy `TestMessageHandler` and reimplement `HttpMessageHandler`.  In other words, write your tests assuming `TestMessageHandler` is a sealed class, it really should be, but that doesn't mean it shouldn't be improved to allow better testing.

Commit:
Removed TestContentTypeMessageHandler.  Added charset check

joe-elliott commented:
Thanks for the feedback.

I removed `TestContentTypeMessageHandler` and added necessary features to `TestMessageHandler`.  I also added a test for charset.

wparad reviewed (approved):
But you'll need to rebase with your other changes before the merge.

Commit:
Merge branch 'master' into cache-content-headers

joe-elliott commented:
Conflicts resolved.
